### PR TITLE
Add optional SELinux support

### PR DESCRIPTION
**Pull request by dburgener**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [x ] All package sources are available
- [x ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

This PR introduces optional SELinux support.  Default Mariner builds will continue to run without SELinux, but the relevant packages are available for those who want to build SELinux enabled Mariner images.

###### Change Log  

- Add the following SELinux packages:
* secilc - low level policy compiler
* mcstrans - usability add ons for Multi-Level Security policies
* libsemanage - policy management capabilities
* selinux-policy - A generic policy from upstream
* checkpolicy - policy compiler
* setools - Policy analysis suite
* policycoreutils - Command line utilities necessary to use SELinux

- Add SELinux support in the follow packages.  These consist of compiler flags to enable SELinux integration in these packages:
* util-linux
* dbus
* systemd
* systemd-bootstrap
* cronie
* coreutils
* openssh
* shadow-utils
* pam
* rpm

- Modify the following packages to address dependency and build issues exposed by above modifications:
* python3
* python2
* curl
* groff

###### Does this affect the toolchain?  

**YES**
NO

###### Associated issues  

- #xxxx

###### Test Methodology

- Validated via pipeline build, and local enablement of SELinux using pipeline artifacts (built image and SELinux packages)

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -33,8 +33,9 @@ What does the PR accomplish, why was it needed?
 
 ###### Does this affect the toolchain?  <!-- REQUIRED -->
 <!-- Any packages which are included in the toolchain should be carefully considered. Make sure the toolchain builds with these changes if so. -->
-**YES**
```
Comment (thcrain-msft) [inline .github/pull_request_template.md:36]: I'm a fan of the changes in this file, but can we get these in a separate PR?

File: SPECS/curl/curl.spec [hunk 2]
```diff
@@ -98,6 +87,9 @@ rm -rf %{buildroot}/*
 %{_libdir}/libcurl.so.*
 
 %changelog
+* Fri May 07 2021 Daniel Burgener <daburgen@microsoft.com> - 7.76.0-2
```
Comment (thcrain-msft) [inline SPECS/curl/curl.spec:90]: Can you give more details on the circular dependency? Asking because recently, in the dev branch, we fixed a problem where our package graph was not taking the `RUN_CHECK` Make variable into account when parsing requirements from spec files.

Link: https://github.com/microsoft/CBL-Mariner/pull/945

Would love to get more details on your problem and see if that fix would work here, since this is a fairly core package to us and it'd be preferable to not remove the check section. It'd also help me prioritize porting this change back to 1.0-dev.

File: .github/pull_request_template.md (hunk 1 above)
Comment (dburgener) [reply .github/pull_request_template.md:36]: That makes a lot of sense.  The origin of these changes came from Daniel McIlvaney, who did a little bit of work on this branch last fall and included these updates in it.  Let me see if I'm able to pull those out for a separate PR, keeping his git info and Sign off.

File: SPECS/curl/curl.spec (hunk 2 above)
Comment (dburgener) [reply SPECS/curl/curl.spec:90]: I've been trying to reproduce this all day with little success.  I suppose it's possible that something has changed upstream already which addresses it.  I'm going to try a full end to end pipeline build with this change removed and see what happens there.

The note I have from my git commit is:

The curl check code causes a circular dependency with shadow-utils (and
    sudo) once the new libsemanage shadow-utils dependency is added (which is
    required for SELinux support).  Comment the check out to see if that
    fixes the issues before further investigation into correct fix.

Unfortunately, that doesn't seem to provide the level of detail that I need to give a perfect answer to your question.  We'll see if an overnight pipeline run gives me a repro and I'll let you know tomorrow.

File: SPECS/cython3/cython3.signatures.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+ "Signatures": {
+  "Cython-0.28.5.tar.gz": "bcd6a24ca0cba54ad8b1755852bb5cd567bb06abea6f776c82fbb10bb46bae37"
```
[the remark this answers is not in the document]
Comment (dburgener) [reply SPECS/cython3/cython3.signatures.json]: I think I removed all the cython additions yesterday prior to this comment.  They don't appear to be in the PR anymore (except in a commit adding them and a later revert).  Am I missing something?

File: SPECS/curl/curl.spec (hunk 2 above)
Comment (dburgener) [reply SPECS/curl/curl.spec:101]: Okay, I was able to get a repro:

ERRO[0035] Unfixable circular dependency of length 142193:	{audit-libs-3.0-6.cm1-RUN<Meta>} --> {audit-libs-3.0-6.cm1-BUILD<Build>} --> {openldap-2.4.57-2.cm1-RUN<Meta>} --> {openldap-2.4.57-2.cm1-BUILD<Build>} --> {cyrus-sasl-2.1.27-4.cm1-RUN<Meta>} --> {systemd-239-37.cm1-RUN<Meta>} --> {systemd-239-37.cm1-BUILD<Build>} --> {cryptsetup-devel-2.3.3-2.cm1-RUN<Meta>} --> {cryptsetup-2.3.3-2.cm1-RUN<Meta>} --> {cryptsetup-libs-2.3.3-2.cm1-RUN<Meta>} --> {cryptsetup-libs-2.3.3-2.cm1-BUILD<Build>} --> {device-mapper-devel-2.03.05-5.cm1-RUN<Meta>} --> {device-mapper-2.03.05-5.cm1-RUN<Meta>} --> {device-mapper-2.03.05-5.cm1-BUILD<Build>} --> {systemd-bootstrap-devel-239-34.cm1-RUN<Meta>} --> {systemd-bootstrap-239-34.cm1-RUN<Meta>} --> {systemd-bootstrap-239-34.cm1-BUILD<Build>} --> {meson-0.56.0-1.cm1-RUN<Meta>} --> {meson-0.56.0-1.cm1-BUILD<Build>} --> {ninja-build-1.8.2-3.cm1-RUN<Meta>} --> {ninja-build-1.8.2-3.cm1-BUILD<Build>} --> {gtest-devel-1.8.1-5.cm1-RUN<Meta>} --> {gtest-devel-1.8.1-5.cm1-BUILD<Build>} --> {cmake-3.17.3-2.cm1-RUN<Meta>} --> {cmake-3.17.3-2.cm1-BUILD<Build>} --> {curl-devel-7.76.0-2.cm1-RUN<Meta>} --> {curl-7.76.0-2.cm1-RUN<Meta>} --> {curl-7.76.0-2.cm1-BUILD<Build>} --> {sudo-1.9.5p2-2.cm1-RUN<Meta>} --> {shadow-utils-4.6-11.cm1-RUN<Meta>} --> {shadow-utils-4.6-11.cm1-BUILD<Build>} --> {libsemanage-devel-2.9-4.cm1-RUN<Meta>} --> {libsemanage(x86-64)-2.9-4.cm1-RUN<Meta>} --> {audit-libs-3.0-6.cm1-RUN<Meta>} 
ERRO[0035] Error found while resolving package dependency cycles: cycle contains build dependencies, unresolvabl

That's just one example.  The full pipeline log is here: https://dev.azure.com/mariner-org/mariner/_build/results?buildId=93957&view=logs&j=75cba72d-09cb-503a-3927-4ab61ac4a179&t=e3d06be3-74ca-556e-743b-b543e06aae03

The core issue as I see it is that libsemanage has a lot of dependencies, and this commit introduces a shadow-utils dependency on libsemanage, which is necessary to set user login contexts with SELinux.  Curl is pulled in by cmake, which is in turn a dependency for a lot of packages.  I'm not sure what the cause of the cmake dependency on curl is and whether there is some way to break the cycle there.  Comparison with Fedora and Red Hat shows that they have no dependency on shadow-utils or sudo in the curl spec file.

If your change would fix this problem retaining the check code, that would be great.

File: .github/pull_request_template.md (hunk 1 above)
Comment (dburgener) [reply .github/pull_request_template.md]: If I've done all the git cleanup correctly, I believe I have now reverted this chunk of changes on this branch and opened a new PR to add them in here: https://github.com/microsoft/CBL-Mariner/pull/963

File: toolkit/scripts/toolchain/build_official_toolchain_rpms.sh [hunk 4]
```diff
@@ -487,6 +485,53 @@ chroot_and_install_rpms gperf
 chroot_and_install_rpms cryptsetup
 build_rpm_in_chroot_no_install systemd
 
+build_rpm_in_chroot_no_install golang-1.15
```
Comment (anphel31) [inline toolkit/scripts/toolchain/build_official_toolchain_rpms.sh:488]: I think this can be changed to simply "golang"

**Comment by anphel31** [comment]
> 

are these additions necessary? This change means every package build chroot will by default include these 6 additional packages.

---
Refers to: toolkit/resources/manifests/package/pkggen_core_x86_64.txt:174 in 77fa250. [](commit_id = 77fa2504b4a4753d2c7c380713cb3e925acf6810, deletion_comment = False)

File: toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go [hunk 5]
```diff
@@ -277,6 +277,7 @@ func (r *RpmRepoCloner) Clone(cloneDeps bool, packagesToClone ...*pkgjson.Packag
 			"--destdir",
 			chrootDownloadDir,
 			pkgName,
+			"--rpmverbosity=yes",
```
Comment (anphel31) [inline toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go:280]: I think this tooling change should be moved into a separate PR; does not appear directly related to SELinux

**Review by anphel31 — APPROVED** [review]

File: toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go (hunk 5 above)
Comment (dburgener) [reply toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go:280]: This change came from when Daniel McIlvaney worked on this branch in the fall.  Now that you highlight it, I'm don't think it's needed at all and can probably be dropped.

**Comment by dburgener** [comment]
@anphel31

> are these additions necessary? This change means every package build chroot will by default include these 6 additional packages.

These changes were added by Daniel McIlvaney when he worked on this branch last fall.  I don't know of any reason why they would be necessary, but I haven't tested without them.  I can do so.

> Can %check sections be added for any of the new SPEC files? Ideally any test assets from the packages that we can call from %check for validation would be helpful. Not blocking for this PR, but would be good to know if tests are available that can be enabled later.

I don't know anything about %check section authoring.  There were not %check sections in the original Fedora spec files that these were based on.  Upstream testing of SELinux functionality is typically done via the SELinux Test Suite (https://github.com/SELinuxProject/selinux-testsuite).  There are a number of setup and environmental constraints in running the test suite that could be an obstacle for implementing via a %check section, but as I mentioned, I'm not sure what all is entailed there.  It would certainly be worth investigating.

On a broad topic, I see you've approved the PR.  You mentioned that the %check section comment was non-blocking, but it does seem like the go tooling and pkggen comments warrant some updates.  Would you like to tackle those updates prior to merge (delaying the merge), or treat them as todos after the fact?

**Comment by dburgener** [comment]
@anphel31 I have kicked off a pipeline run with the pkggen lines removed and the go tooling change reverted, so we will see in about 16 hours (or less) what the impact of those changes is.
